src/main/index.ts

```diff
@@ -43,7 +49,13 @@ app.on('ready', async () => {
     await installDevToolsExtensions();
   }
   await createWindow();
+
+  const settings = readSettingsFile();
+  if (settings) {
+    mainStore.dispatch(setAllSettings(settings));
+  }
   mainStore.dispatch(setVersionDetails(getVersionDetails()));
+  observeStore(mainStore, selectAllSettings, writeSettingsFile);
```

Comment:
The only gripe I have with this is that over time we might build up the amount of observeStores we have. I tried to play around with the actions fired on main thread and there might be a way to actually use them like we'd want.

If we'd change the `persistSettings` functions to accept the path as a param:

```ts
const settingsFileName = 'publab-settings.json';

export function writeSettingsFile(path: string, settings: Settings) {
  const settingJSON = JSON.stringify(settings, null, 2);
  writeFileSync(`${path}/${settingsFileName}`, settingJSON);
}

export function readSettingsFile(path: string): Settings | null {
  const fullPath = `${path}/${settingsFileName}`;
  if (existsSync(fullPath)) {
    const settingsJSON = readFileSync(`${path}/${settingsFileName}`).toString();
    return JSON.parse(settingsJSON);
  }
  return null;
}
```

and then created the action with `createAsyncActionMain` helper within the `settingsSlice`:

```ts
export const saveSettingsThunk = createAsyncActionMain<{
  path: string;
  settings: Settings;
}>('saveSettings', ({ path, settings }) => async (dispatch) => {
  dispatch(setAllSettings(settings));

  writeSettingsFile(path, settings);
});
```

we would then be able to fire it from the renderer code by getting the `app` from remote module:

```tsx
const path = require('electron').remote.app.getPath('userData');

const AppSettings = () => {
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const [settings, changeSetting] = useReducer(
    handleSettingChange,
    useSelector(selectAllSettings)
  );

  function submitChanges() {
    dispatch(saveSettingsThunk({ path, settings }));
  }

  return (
    <div>
      // ... other code
      <Button onClick={() => submitChanges()}>{t('common.save')}</Button>
    </div>
  );
};
```

I feel like this is more inline with the idea behind using [electron-redux](https://github.com/klarna/electron-redux) package. It also means that the code, which performs the save of the settings is fired by the button press rather than by listening to changes of the store value.

What do you think @LostInStatic?

> We could also potentially create a set of `renderer/constants.ts` for things like this user path that could only be used in front-end code since they would utilise the remote package to get the app.

Reply:
The biggest issue I had with your proposed solution was the shift in responsibility for "knowing" the path to the UI component. However, this gave me a hunch about another solution to the problem. A module like this would allow us to safely (as both have the same type) reference `app`  in the code shared between processes (in this case, settings slice):
``` javascript
import { app as mainApp, remote } from 'electron';

const isRenderer = process && process.type === 'renderer';

const app = isRenderer ? remote.app : mainApp;
export default app;
```
Did a quick test, it seems to work as intended. 

I am still not fully convinced against the observer in this specific case. On one hand, I also don't like the idea of constantly monitoring the state for changes (especially taking into account the potential issues with deep comparison). On the other, handling in in the reducer means having to remember about saving the settings in every reducer or using only `setAllSettings` to handle every change. 

Taking all into consideration, I'm the least doubtful (even if by a small margin) about the solution involving calling `writeSettingsFile`  in every reducer - hopefully we will not need too many ; )

As that is he approach you also leaned towards I'll implement that, including the module described above.

Reply:
@LostInStatic let'd do as you say for now. 👍 

I think we should have a brainstorm at the end of the year to evaluate the architecture of the app and whether `elecron-redux`'s aliased actions that can make code run on the main thread are actually helping us or weirdly bluring the border between responsibilities of renderer and main threads.

Reply:
Added two thunks to `settingsSlice` (file input and output), made changes (now takes partial settings object and spreads it with old state) to `setAllSettings` to cover most practical needs and abstracted file operations (so it can be reused) - now most information directly related to the app settings is contained in the slice.

I went ahead and converted the existing code to use new thunks, removed unnecessary exports and reducers. I also moved version details out of the slice (as there is a convenient way to access the information directly).